### ISSUE TITLE
REPL prompt hooks need to be adjusted for brave new world

### ISSUE THREAD

**Issue by Keno** [JuliaLang/julia#19809]
These hooks: https://github.com/JuliaLang/julia/blob/c38a5a3044ffd62c585225a201422f95fd18f6dc/base/LineEdit.jl#L629-L630
need to be adjusted for the world counter. @vtjnash says the semantics of this will change before the release, so nothing to do for now, but we should revisit this before we're done with the release.

**Comment by StefanKarpinski** [comment]
>  @vtjnash says the semantics of this will change before the release

Go on... how so?

**Comment by Keno** [comment]
If I knew I'd have included that information in the issue ;)

**Comment by luchr** [comment]
Here is an ugly workaround (for the lines mentioned):
```julia
prefix = isa(p.prompt_prefix,Function) ? eval(Expr(:call,p.prompt_prefix)) : p.prompt_prefix
suffix = isa(p.prompt_suffix,Function) ? eval(Expr(:call,p.prompt_suffix)) : p.prompt_suffix
```
But there has to be a better solution "capturing" the type of the Functions (via a parametrized methods), so that this dynamic runtime dispatch can be eliminated at compile-time; I hope.

**Comment by carlobaldassi** [comment]
I guess that the following issue is also related: I can no longer redefine the keymap as I used to. Consider the following simplified `.juliarc.jl`:

```julia
const mykeys = Dict{Any,Any}(
    # Up Arrow
    "\e[A" => (s,o...)->nothing,
)

function myrepl(repl)
    repl.interface = Base.REPL.setup_interface(repl; extra_repl_keymap = mykeys)
    return
end

atreplinit(myrepl)
```

With this, pressing the up arrow gives:

```
julia> WARNING: Caught an exception in the keymap:
ERROR: MethodError: no method matching (::##1#2)(::Base.LineEdit.MIState, ::Base.REPL.LineEditREPL, ::String)
The applicable method may be too new: running in world age 20628, while current world is 20630.
Closest candidates are:
  #1(::Any, ::Any...) at /home/carlo/.juliarc.jl:3 (method too new to be called from this world context.)
```

and I have no idea how to fix this at the moment, everything I tried failed with the same message.

**Comment by luchr** [comment]
@carlobaldassi  That's issue #19718.

**Comment by carlobaldassi** [comment]
ah yes thanks @luchr

**Comment by tkelman** [comment]
was this fixed by #19924 ? cc @KristofferC

**Comment by StefanKarpinski** [comment]
The specific problem with the REPL is fixed. @vtjnash may add a way of doing this more conveniently, but that can happen when he gets to it (so called "callback wrapper" feature).
